//--- a/target/linux/ath79/generic/config-defaultx
//+++ b/target/linux/ath79/generic/config-defaultx
@@ -16,7 +16,8 @@
 CONFIG_MTD_REDBOOT_PARTS=y
 CONFIG_MTD_SPI_NOR_SWP_DISABLE=y
 # CONFIG_MTD_SPI_NOR_SWP_DISABLE_ON_VOLATILE is not set
-CONFIG_MTD_SPI_NOR_USE_VARIABLE_ERASE=y
+CONFIG_MTD_SPI_NOR_USE_4K_SECTORS=y
+CONFIG_MTD_SPI_NOR_USE_VARIABLE_ERASE=n
 CONFIG_MTD_SPLIT_EVA_FW=y
 CONFIG_NVMEM_SYSFS=y
 CONFIG_NVMEM_U_BOOT_ENV=y
--- /dev/null
+++ b/target/linux/generic/backport-6.6/420-earlier-mtd-setup.patch
@@ -0,0 +1,20 @@
+--- a/drivers/mtd/spi-nor/core.c
++++ b/drivers/mtd/spi-nor/core.c
+@@ -3540,14 +3540,14 @@
+ 	if (ret)
+ 		return ret;
+ 
++	/* No mtd_info fields should be used up to this point. */
++	spi_nor_set_mtd_info(nor);
++
+ 	/* Send all the required SPI flash commands to initialize device */
+ 	ret = spi_nor_init(nor);
+ 	if (ret)
+ 		return ret;
+ 
+-	/* No mtd_info fields should be used up to this point. */
+-	spi_nor_set_mtd_info(nor);
+-
+ 	dev_info(dev, "%s (%lld Kbytes)\n", info->name,
+ 			(long long)mtd->size >> 10);
+ 
--- /dev/null
+++ b/target/linux/generic/backport-6.6/420-flash-fixes.patch
@@ -0,0 +1,10 @@
+--- a/drivers/mtd/spi-nor/spansion.c
++++ b/drivers/mtd/spi-nor/spansion.c
+@@ -826,6 +826,7 @@
+ 		NO_SFDP_FLAGS(SECT_4K | SPI_NOR_DUAL_READ |
+ 			      SPI_NOR_QUAD_READ) },
+ 	{ "s25fl064k",  INFO(0xef4017,      0,  64 * 1024, 128)
++		FLAGS(SPI_NOR_HAS_LOCK)
+ 		NO_SFDP_FLAGS(SECT_4K | SPI_NOR_DUAL_READ |
+ 			      SPI_NOR_QUAD_READ) },
+ 	{ "s25fl116k",  INFO(0x014015,      0,  64 * 1024,  32)
--- /dev/null
+++ b/target/linux/ath79/patches-6.6/941-mikrotik-nand-flash.patch
@@ -0,0 +1,11 @@
+--- a/drivers/mtd/nand/raw/nand_amd.c
++++ b/drivers/mtd/nand/raw/nand_amd.c
+@@ -44,6 +44,8 @@
+ 		chip->options |= NAND_BBM_FIRSTPAGE | NAND_BBM_SECONDPAGE |
+ 				 NAND_BBM_LASTPAGE;
+ 
++ 	chip->options |= NAND_NO_SUBPAGE_WRITE;
++
+ 	return 0;
+ }
+ 
